File: SocialChallenge/Utilities/OperationCoordinator/JSCOperationCoordinator.m
//
//  JSCOperationCoordinator.m
//  SocialChallenge
//
//  Created by Javier Laguna on 09/08/2015.
//
//

#import "JSCOperationCoordinator.h"

#import "JSCOperation.h"

NSString * const kJSCNetworkDataOperationSchedulerTypeIdentifier = @"kJSCNetworkDataOperationSchedulerTypeIdentifier";

NSString * const kJSCLocalDataOperationSchedulerTypeIdentifier = @"kJSCLocalDataOperationSchedulerTypeIdentifier";

@interface JSCOperationCoordinator ()

/**
 Contains the schedulers in the app.
 */
@property (nonatomic, strong) NSMutableDictionary *mutableSchedulerTable;

/**
 Finds the first operation which can coalesce the passed in operation
 and coalesce with it.
 
 @param operation - opearation to coaleque with.
 @param queue - queue to coalesce the operation on.
 
 @return the coalesced operation
 */
- (JSCOperation *)coalesceOperation:(JSCOperation *)newOperation
                              queue:(NSOperationQueue *)queue;

@end

@implementation JSCOperationCoordinator

#pragma mark - Singleton

+ (instancetype)sharedInstance
{
    static JSCOperationCoordinator *sharedInstance = nil;
    
    static dispatch_once_t onceToken;
    
    dispatch_once(&onceToken, ^
                  {
                      sharedInstance = [[JSCOperationCoordinator alloc] init];
                  });
    
    return sharedInstance;
}

#pragma mark - Init

- (instancetype)init
{
    self = [super init];
    
    if (self)
    {
        self.mutableSchedulerTable = [[NSMutableDictionary alloc] init];
    }
    
    return self;
}

#pragma mark - Register

- (void)registerQueue:(NSOperationQueue *)queue
  schedulerIdentifier:(NSString *)schedulerIdentifier
{
    self.mutableSchedulerTable[schedulerIdentifier] = queue;
}

#pragma mark - Add

- (void)addOperation:(JSCOperation *)operation
{
    NSOperationQueue *queue = self.mutableSchedulerTable[operation.targetSchedulerIdentifier];
    
    JSCOperation *coalescedOperation = [self coalesceOperation:operation
                                                         queue:queue];
    
    if (!coalescedOperation)
    {
        [queue addOperation:operation];
    }
}

#pragma mark - Coalescing

- (JSCOperation *)coalesceOperation:(JSCOperation *)newOperation
                              queue:(NSOperationQueue *)queue
{
    NSArray *operations = [queue operations];
    
    for (JSCOperation *operation in operations)
    {
        BOOL canAskToCoalesce = [operation isKindOfClass:[JSCOperation class]];
        
        if (canAskToCoalesce &&
            [operation canCoalesceWithOperation:newOperation])
        {
            [operation coalesceWithOperation:newOperation];
            return operation;
        }
    }
    
    return nil;
}

@end

File: SocialChallenge/Utilities/OperationCoordinator/JSCOperationCoordinator.swift
//
//  JSCOperationCoordinator.swift
//  SocialChallenge
//
//  Created by Javier Laguna on 16/03/2016.
//
//

import Foundation

/**
Identifier for network based operations.
*/
let kJSCNetworkDataOperationSchedulerTypeIdentifier = "kJSCNetworkDataOperationSchedulerTypeIdentifier"

/**
Identifier for local operations.
*/
let kJSCLocalDataOperationSchedulerTypeIdentifier = "kJSCLocalDataOperationSchedulerTypeIdentifier"

/**
 This class handles the schedulers that run the operations.
 */
class JSCOperationCoordinator {
    
    /**
     Contains the schedulers in the app.
     */
    lazy var schedulerDictionary: Dictionary <String, NSOperationQueue> = {
        
        return [:]
    }()
    
    //MARK: Singleton
    
    static let sharedInstance = JSCOperationCoordinator()
    
    //MARK: Register
    
    /**
    Registers a Scheduler with the coordinator.
    
    - Parameter queue -  Scheduler to register with the coordinator.
    - Parameter schedulerIdentifier Scheduler identifier to register the coordinator under.
    */
    func registerScheduler(queue: NSOperationQueue, schedulerIdentifier: String) {
        
        self.schedulerDictionary[schedulerIdentifier] = queue
    }
    
    //MARK: Add
    
    /**
    Adds operation to the scheduler, if an operation with the same identifier is on the scheduler then the coordinator
    will perform coalescing.
    
    - Parameter operation - operation to add.
    */
    func addOperation(operation: JSCOperation) {
        
        let queue = self.schedulerDictionary[operation.targetSchedulerIdentifier!]! as NSOperationQueue
        
        let coalescedOperation = self.coalesceOperation(operation, queue: queue)
        
        if coalescedOperation == nil {
            
            queue.addOperation(operation)
        }
    }
    
    //MARK: Coalescing
    
    /**
    Finds the first operation which can coalesce the passed in operation
    and coalesce with it.
    
    - Parameter operation - opearation to coaleque with.
    - Parameter scheduler - Scheduler to coalesce the operation on.
    
    - Returns the coalesced operation
    */
    func coalesceOperation(newOperation: JSCOperation, queue: NSOperationQueue) -> JSCOperation? {
        
        let operations = queue.operations
        
        for operation in operations as! Array <JSCOperation> {
            
            let canAskToCoalesce = operation.isKindOfClass(JSCOperation.self) as Bool
            
            if canAskToCoalesce &&
                operation.canCoalesceWithOperation(newOperation) {
                    
                    operation.coalesceWithOperation(newOperation)
                    
                    return operation;
            }
        }
        
        return nil;
    }
}